### ISSUE TITLE
Document structure is not preserved when doing xpath on XML object

### ISSUE THREAD

**Issue by skapral** [jcabi/jcabi-xml#131]
Attached tests demonstrating the issue

[proba-xml.tar.gz](https://github.com/jcabi/jcabi-xml/files/863055/proba-xml.tar.gz)

It seems like every call on XML::nodes method creates an independent version(s) of document(s) with root(s) on the node(s) found, which in turn makes navigation to parent nodes impossible.

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this issue

**Comment by yegor256** [comment]
@skapral with your permission I'll post it here:

```
package com.proba;

import com.jcabi.xml.XML;
import com.jcabi.xml.XMLDocument;
import java.io.ByteArrayInputStream;
import java.util.List;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;
import static org.junit.Assert.*;
import org.junit.Test;
import org.w3c.dom.Document;
import org.w3c.dom.Node;

public class XpathTest {
    @Test
    public void xpathWithDom() throws Exception {
        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
        Document doc = documentBuilder.parse(new ByteArrayInputStream("<base><item1/><item2/><item3/></base>".getBytes()));
    
        XPathFactory xpathFactory = XPathFactory.newInstance();
        XPath xpath = xpathFactory.newXPath();
        Node item2 = (Node) xpath.compile("//base/item2").evaluate(doc, XPathConstants.NODE);
        Node parent = (Node) xpath.compile("..").evaluate(item2, XPathConstants.NODE);
        
        assertNotNull(parent);
        assertEquals("base", parent.getNodeName());
    }
    
    @Test
    public void xpathWithJcabi() {
        XML xml = new XMLDocument("<base><item1/><item2/><item3/></base>");
        XML item2 = xml.nodes("//base/item2").get(0);
        XML parent = item2.nodes("..").get(0);
        
        assertNotNull(parent);
        assertEquals("base", parent.node().getNodeName());
    }
}

```

**Comment by yegor256** [comment]
@skapral this is _by design_ I guess. This is what happens in your code. First, we [fetch](https://github.com/jcabi/jcabi-xml/blob/0.18.1/src/main/java/com/jcabi/xml/XMLDocument.java#L429):

```
xpath.evaluate(query, this.node(), qname);
```

Then, we [create](https://github.com/jcabi/jcabi-xml/blob/0.18.1/src/main/java/com/jcabi/xml/XMLDocument.java#L356-L360) new _immutable_ objects of class `XMLDocument`:

```
new XMLDocument(nodes.item(idx), this.context, true)
```

Then, we [serialize it](https://github.com/jcabi/jcabi-xml/blob/0.18.1/src/main/java/com/jcabi/xml/XMLDocument.java#L272) in order to encapsulate: 

```
this.xml = XMLDocument.asString(node);
```

In other words, every instance of `XMLDocument` encapsulates a `String`, not a `Node`. That's why it doesn't know anything about the parent. This is done in order to make `XMLDocument` immutable. This was an important goal of this library -- to make sure all objects are immutable.

**Comment by skapral** [comment]
@yegor256 If it is all about immutability, then I guess introducing method XML::node was a dangerous idea initially. According to its javadoc, it is supposed to unwrap the wrapped node, which is stated in its documentation:
>     * Retrieve DOM node, represented by this wrapper.
>     * @return DOM node

In reality, this statement breaks after applying any XPath on XML. XML:nodes javadoc states that it:
>      * Retrieve DOM nodes from the XML response.

And it is not the whole truth: nobody tells that besides extracting these nodes, it converts them to separate document. XML instances produced by XmlDocument::nodes doesn't do what described in documentation, and I believe it is a problem.

In my pull request https://github.com/jcabi/jcabi-xml/pull/132 I did changes assuming that the class encapsulates the Node and not its contents - seems like wrong approach. But I think that the fact that XMLDocument encapsulates String and not the Node doesn't mean that there is no ability to introduce XPath navigation in both sides. What if we intorduce other XML implementation which encapsulates exactly the Node found?

I can guess objections like it may be possible to mutate such implementation through XML::node. I agree, and that's why I initially said that introducing XML::node was a dangerous idea. Or maybe not? What do you think?

**Comment by skapral** [comment]
The previous message is rather long, so I decided to outline it to separate issue. In scope of this one, what do you think about this statement?
> What if we introduce other XML implementation which encapsulates exactly the Node found, and use it in XmlDocument::nodes instead of XMLDocument?

**Comment by yegor256** [comment]
@skapral yes, I like the idea of introducing another `XML` implementation, which will behave differently. But don't forget to change the documentation at the `XML` interface.

**Comment by skapral** [comment]
After looking at the issue with fresh mind, I found more compact solution. See the PR.

**Comment by 0crat** [comment]
Job `gh:jcabi/jcabi-xml#131` is not assigned, can't get start time

**Comment by 0crat** [comment]
Job [#131](https://github.com/jcabi/jcabi-xml/issues/131) was not in scope
